capabilities/writetarget/write_target.go

```diff
@@ -0,0 +1,505 @@
+}
+
+// TODO: opts.Config input is not validated for sanity
+func NewWriteTarget(opts WriteTargetOpts) capabilities.ExecutableCapability {
+	capInfo := capabilities.MustNewCapabilityInfo(opts.ID, capabilities.CapabilityTypeTarget, CapabilityName)
+
+	// override Unknown to ensure Finalized is default
+	if opts.WriteAcceptanceState == 0 {
+		opts.WriteAcceptanceState = commontypes.Finalized
+	}
+
+	return &writeTarget{
+		capInfo,
+		opts.Config,
+		opts.ChainInfo,
+		opts.Logger,
+		opts.Beholder,
+		opts.ChainService,
+		opts.ConfigValidateFn,
+		opts.NodeAddress,
+		opts.ForwarderAddress,
+		opts.TargetStrategy,
+		opts.WriteAcceptanceState,
+	}
+}
+
+func success() capabilities.CapabilityResponse {
+	return capabilities.CapabilityResponse{}
+}
+
+func (c *writeTarget) Execute(ctx context.Context, request capabilities.CapabilityRequest) (capabilities.CapabilityResponse, error) {
+	// Take the local timestamp
+	tsStart := time.Now().UnixMilli()
+
+	// Trace the execution
+	attrs := c.traceAttributes(request.Metadata.WorkflowExecutionID)
+	ctx, span := c.beholder.Tracer.Start(ctx, "Execute", trace.WithAttributes(attrs...))
+	defer span.End()
+
+	// Notice: error skipped as implementation always returns nil
+	capInfo, _ := c.Info(ctx)
+
+	c.lggr.Debugw("Execute", "request", request, "capInfo", capInfo)
+
+	// Helper to keep track of the request info
+	info := &requestInfo{
+		tsStart:   tsStart,
+		node:      c.nodeAddress,
+		forwarder: c.forwarderAddress,
+		receiver:  "N/A",
+		request:   request,
+		reportInfo: &reportInfo{
+			reportContext: nil,
+			report:        nil,
+			signersNum:    0, // N/A
+			reportID:      0, // N/A
+		},
+		reportTransmissionState: nil,
+	}
+	// Helper to build monitoring (Beholder) messages
+	builder := NewMessageBuilder(c.chainInfo, capInfo)
+
+	// Validate the config
+	receiver, err := c.configValidateFn(request)
+	if err != nil {
+		msg := builder.buildWriteError(info, 0, "failed to validate config", err.Error())
+		return capabilities.CapabilityResponse{}, c.asEmittedError(ctx, msg)
+	}
+
+	// Source the receiver address from the config
+	info.receiver = receiver
+
+	// Source the signed report from the request
+	signedReport, ok := request.Inputs.Underlying[KeySignedReport]
+	if !ok {
+		cause := fmt.Sprintf("input missing required field: '%s'", KeySignedReport)
+		msg := builder.buildWriteError(info, 0, "failed to source the signed report", cause)
+		return capabilities.CapabilityResponse{}, c.asEmittedError(ctx, msg)
+	}
+
+	// Decode the signed report
+	inputs := types.SignedReport{}
+	if err = signedReport.UnwrapTo(&inputs); err != nil {
+		msg := builder.buildWriteError(info, 0, "failed to parse signed report", err.Error())
+		return capabilities.CapabilityResponse{}, c.asEmittedError(ctx, msg)
+	}
+
+	// Source the report ID from the input
+	info.reportInfo.reportID = binary.BigEndian.Uint16(inputs.ID)
+
+	err = c.beholder.ProtoEmitter.EmitWithLog(ctx, builder.buildWriteInitiated(info))
```

Comment:
nit: Should this metric be emited only after verification that this was a non-empty report?
Meaning, move this emit call to after line 294?
Or does WriteInitiated just imply it was initiated, but still to check if report is valid etc etc.

Reply:
Or maybe we move this emit to line 328, once we determine we actually need to send a writeRequet.

Reply:
The `WriteInitiated` event represents a signal that the WT.Execute was called and metadata/timings of the call.

This is the immediate event on fn trigger, before any validations, before any (buisiness logic) errors, before anything happens we want to get an event that `WriteInitiated`.

Our dashboards and SLA metrics are based on that event.